### PR TITLE
Ignore resources and instances that are already FULLAUTO in HelixBootstrapUpgradeTool

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Ignore any updates to resource and instances that are aleady in FULL AUTO in HelixBootstrapUpgradeTool

### CODE REVIEW

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java [hunk 1]
```diff
@@ -1340,6 +1361,12 @@ private void updateClusterMapInHelix(boolean startValidatingClusterManager) thro
     }
     // make sure bootstrap has completed in all dcs (can extend timeout if amount of resources in each datacenter is really large)
     bootstrapLatch.await(15, TimeUnit.MINUTES);
+    for (Map.Entry<String, Exception> entry : exceptions.entrySet()) {
+      if (entry.getValue() != null) {
```
Comment (JingQianCloud) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:1365]: If we have multiple entries with exception, we'll throw exception and leave the function. Won't be able to print all the warnings.
Is this expected?

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java [hunk 2]
```diff
@@ -1515,19 +1542,22 @@ private void updateIdealStateWithNewPartitionsForFullAuto(String dcName,
...
           // Partition doesn't belong to any host that are already in the helix
         } else if (resources.size() == 1) {
+          info("[{}] Partition {} belongs to resource {} at round {}", dcName, ent.getKey(),
+              resources.iterator().next(), round);
           allPartitionsDontHaveResources = false;
           int resourceId = resources.iterator().next();
-          resourceIdToInstances.get(resourceId).addAll(instanceNames);
           // Some instance names might not have a resource name, now assign the resource name to it
-          instanceNames.forEach(in -> instanceNameToOneResource.put(in, resourceId));
+          instanceNames.forEach(in -> instanceNameToOneResource.put(in + "_" + numReplica, resourceId));
```
Comment (JingQianCloud) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:1559]: Is there any hard rule or naming convention for the "resource"?

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java [hunk 3]
```diff
@@ -1605,86 +1640,134 @@ private void updateIdealStateWithNewPartitionsForFullAuto(String dcName,
...
-            info("[{}] Update resource {} in datacenter {} from {} partitions to {} partitions", dcName.toUpperCase(),
-                resourceName, dcName, sizeBefore, partitionsUnderResource.size());
-          } else {
-            info("[{}] Under DryRun mode, update resource {} in datacenter {} from {} partitions to {} partitions",
-                dcName.toUpperCase(), resourceName, dcName, sizeBefore, partitionsUnderResource.size());
-          }
-          resourcesUpdated.incrementAndGet();
-        } else {
-          info("[{}] No changes made to resource {}, skip", dcName.toUpperCase(), resourceName);
+    for (Set<String> instances : newGroups) {
```
Comment (JingQianCloud) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:1672]: What's the "group" concept in the Helix concept?

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java [hunk 4]
```diff
@@ -2625,6 +2722,118 @@ private void verifyDataNodeAndDiskEquivalencyInDc(Datacenter dc, String clusterN
...
+  /**
+   * Fetch IdealStates for all the resource ids.
+   * @param admin The helix admin object.
+   * @param dcName The dc name.
+   * @return
+   */
+  Map<Integer, IdealState> getIdealStateForAllResource(HelixAdmin admin, String dcName) {
+    Map<Integer, IdealState> allIdealStates = new HashMap<>();
+    for (String resourceName : admin.getResourcesInCluster(clusterName)) {
+      if (!resourceName.matches("\\d+")) {
```
Comment (JingQianCloud) [inline ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:2734]: What does that mean? Is \d some special character for us?

**Review by JingQianCloud — COMMENTED** [review]
I don't have much context of the code. 
I'll reviewer as a "watcher". 

Have some questions embedded.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java (hunk 1 above)
Comment (justinlin-linkedin) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:1365]: that's fine, multiple exceptions can be resolved one at a time.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java (hunk 2 above)
Comment (justinlin-linkedin) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:1559]: resource names are going to be integer numbers.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java (hunk 3 above)
Comment (justinlin-linkedin) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:1669]: Groups means a set of instances that share partitions with each other. Instances in one group should never share partitions with any instances in other groups.

File: ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java (hunk 4 above)
Comment (justinlin-linkedin) [reply ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java:2733]: it just means that resources has to be a number.

**Review by Arun-LinkedIn — APPROVED** [review]

**Review by JingQianCloud — APPROVED** [review]

**Pull request merged**